conn_native.go

```diff
@@ -113,10 +121,18 @@ func (c *Conn) SetDeadline(t time.Time) error {
 }
 
 func (c *Conn) SetReadDeadline(t time.Time) error {
+	// Don't lock when setting the read deadline. That would prevent us from
+	// interrupting an in-progress read.
 	return c.Conn.SetReadDeadline(t)
 }
 
 func (c *Conn) SetWriteDeadline(t time.Time) error {
+	// Unlike the read deadline, we need to lock when setting the write
+	// deadline.
+
+	c.writeLock.Lock()
+	defer c.writeLock.Unlock()
```

Comment:
I recommend unifying these into a `sync.RWMutex` and use `lock.RLock()` in combination with `lock.Lock()`. Since you set the write lock without checking the read lock.

Reply:
We're locking reads independently from writes. The underlying library does not allow concurrent reads from multiple threads or concurrent writes from multiple threads.

However, reading while writing (as long as only one thread is reading reading and only one thread is writing) is perfectly fine. As a matter of fact, that's how we normally use connections.

Note: the _specific_ bug we were fixing here was concurrently setting a deadline while writing. We _could_ have simply locked the write side and that's it, but we've run into hard to track down race conditions in the past where we were reading from multiple threads without synchronizing (usually in error paths where we're throwing away the connection anyways).

Reply:
Got it! Thank you for the explanation!